torch/csrc/jit/passes/integer_value_refinement.cpp

```diff
@@ -0,0 +1,199 @@
+    active_refinements_.push_back(&block_refinements);
+    for (Node* n : b->nodes()) {
+      if (n->matches("aten::eq(int a, int b) -> bool") ||
+          n->matches("aten::ne(int a, int b) -> bool")) {
+        for (size_t const_index : {0, 1}) {
+          if (auto ival = constant_as<int64_t>(n->inputs().at(const_index))) {
+            IntegerRefinement refine;
+            refine[n->inputs().at(1 - const_index)] = *ival;
+            info_[n->output()] = n->kind() == aten::eq
+                ? BooleanRefinementMapping::TrueRefinements(std::move(refine))
+                : BooleanRefinementMapping::FalseRefinements(std::move(refine));
+          }
+        }
+      }
+      if (n->kind() == prim::RaiseException) {
+        throwing_blocks_.insert(b);
+      }
+
+      for (size_t input = 0; input < n->inputs().size(); ++input) {
+        Value* input_v = n->inputs().at(input);
+        if (!input_v->type()->cast<IntType>()) {
+          continue;
+        }
+
+        if (auto refine = tryFindRefinement(input_v)) {
+          WithInsertPoint guard(n);
+          auto refine_constant =
+              graph_->insertConstant(static_cast<int64_t>(*refine));
+          n->replaceInputWith(input_v, refine_constant);
+          changed_ = true;
+        }
+      }
+
+      if (n->kind() == prim::If) {
+        IfView if_n(n);
+        bool has_cond_ref = info_.count(if_n.cond()) != 0;
+        ListRefinement empty;
+        auto true_block_refinements = RefineListLens(
+            if_n.thenBlock(),
+            has_cond_ref ? info_[if_n.cond()].true_refine() : empty);
+        auto false_block_refinements = RefineListLens(
+            if_n.elseBlock(),
+            has_cond_ref ? info_[if_n.cond()].false_refine() : empty);
+
+        // we are looking for cases where we can replace
+        // both block outputs with the same value, which opens up
+        // further optimization opportunities
+        // The pass will already handle if both are refined to the same
+        // constant. Here, we add a case case where one block output
+        // is refined to a constant in the other block, and where the existing
+        // block output in the other block is the same constant
+        // x = 1
+        // if y == 1:
+        //    return x
+        // else:
+        //    return y
+        // can always safely be replaced with `y`
+        // this is an important case for symbolic shape analysis
+
+        for (size_t block_index : {0, 1}) {
+          Block* if_block = if_n.node()->blocks().at(block_index);
+          Block* other_if_block = if_n.node()->blocks().at(1 - block_index);
+          for (size_t i = 0; i < n->outputs().size(); ++i) {
+            Value* block_output = if_block->outputs().at(i);
+            if (!block_output->type()->cast<IntType>()) {
+              continue;
+            }
+            // Value must be in scope for both blocks
+            if (!if_n.node()->isDominatedBy(block_output->node())) {
```

Comment:
`if_n` is the if-node, `block_output` is `if_block->outputs().at(i)` where `if_block` is `if_n.node()->blocks().at(block_index)`. Isn't `if_n` always dominated by `block_output`? Adding an example would help here.

Reply:
the one block output doesn't necessarily dominate the other block output

Reply:
I see. I thought if-node output is always a phi-node that merges values from both branches thus would always dominate the if-node. People may get confused if they do not know the IR structure intimately, adding an example would help a lot here.